File: src/main/java/SquareGraph.java
import java.util.stream.IntStream;

public class SquareGraph {
    final int number;

    private int[][] edges;

    public SquareGraph(int number) {
        this.number = number;
        edges = new int[number][];
        IntStream.range(0, number).forEach(vertex -> {
            edges[vertex] = IntStream
                    .rangeClosed(1, number)
                    .map(v -> number - v)
                    .filter(v -> v != vertex)
                    .filter(v -> isSquare(vertex, v))
                    .map(v -> v + 1)
                    .toArray();
        });
    }

    private static boolean isSquare(int vertex, int candidate) {
        return Edges.isSquare(vertex + 1, candidate + 1);
    }

    public int[] getNeighbors(int vertex) {
        return edges[vertex - 1];
    }

    public int degree(int vertex) {
        return edges[vertex - 1].length;
    }
}

File: src/main/java/search/SquareGraph.kt
package search

import util.SquareCache

class SquareGraph(val number: Int, squareCache: SquareCache = SquareCache(2 * number)) {

    private val edges: Array<IntArray>

    init {
        val squares = (1 until 2 * number).filter(squareCache::contains)
        edges = Array(number) { index ->
            val startNode = index + 1
            val neighbors = mutableListOf<Int>()
            squares
                    .filter { it != startNode }
                    .forEach { s ->
                        val endNode = s - startNode
                        if (endNode in 1..number) neighbors += endNode
                    }
            neighbors.sortedDescending().toIntArray()
        }
    }

    fun getNeighbors(vertex: Int): IntArray = edges[vertex - 1]

    fun degree(vertex: Int): Int = edges[vertex - 1].size
}
